src/main/java/com/atlassian/db/replica/spi/circuitbreaker/CircuitBreaker.java

```diff
@@ -1,9 +1,18 @@
 package com.atlassian.db.replica.spi.circuitbreaker;
 
-import com.atlassian.db.replica.api.circuitbreaker.BreakerState;
+import com.atlassian.db.replica.api.SqlCall;
+
+import java.sql.SQLException;
 
 public interface CircuitBreaker {
-    BreakerState getState();
 
-    void handle(Throwable throwable);
+    boolean canCall();
+
+    <T> T handle(SqlCall<T> call) throws SQLException;
```

Comment:
AFAIK we can use `DualCall` to handle all calls to the DB. Maybe,  `CircuitBreaker` interface can contain only `boolean canCall();` method?

Reply:
`DualCall` is about splitting behavior across main/replica
`CircuitBreaker` is about protecting any usage of `DualConnection`, regardless of main/replica

Reply:
`DualCall` is about intercepting calls to main and replica. It can be used to add logs, monitoring or feed CircuitBreaker with better granularity than the current SPI of CB ("entities should not be multiplied without necessity"). It probably can and will be used in the future for use cases we can't imagine now.

Reply:
`BreakerConnection` doesn't use it. Should it? But `BreakerConnection` doesn't split know about main/replica.

Reply:
> BreakerConnection doesn't use it. Should it?

IMO it should use it. `BreakerConnection` was created with `MissingCoverageBreaker` implementation in mind, which doesn't need the split. But it's a special case. In general, you probably want to detect when the replica fails. If only connections to the main are failing, then there's no reason to switch all the traffic to the main database.

If we add an extra entity (CB#handle), it will be more limited than the current SPI `DualCall`.

Reply:
I understand

- [ ] remake `BreakerConnection` to handle replica calls only

Reply:
I'm not sure if we are on the same page. `BreakerConnection` can handle both calls,  how to implement `MissingCoverageBreaker` with replica calls only?

Reply:
Currently, we have two ways to do the same. One gives us a more granular view of what happens (splits calls), another gives us only an aggregation. I'd leave more powerful way instead of:
- adding an alternative way to do the same but with limits (aggregation),
- limiting CB only to get part of the calls.

Reply:
Please check this PR https://github.com/dagguh/db-replica-1/pull/1